firedrake/mesh.py

```diff
@@ -1537,6 +1686,175 @@ def ExtrudedMesh(mesh, layers, layer_height=None, extrusion_type='uniform', kern
+    straight edges. If not, the particles may be placed in the wrong
+    cells.
+
+    :arg plex: the DMPlex within with the DMSwarm should be
+        immersed.
+    :arg coords: an `ndarray` of (npoints, coordsdim) shape.
+    :kwarg fields: An optional list of named data which can be stored
+        for each point in the DMSwarm. The format should be
+        `[(fieldname1, blocksize1, dtype1),
+          ...,
+         (fieldnameN, blocksizeN, dtypeN)]`.
+        For example, the swarm coordinates themselves are stored in a
+        field named `DMSwarmPIC_coor` which, were in not created
+        automatically, would be initialised with
+        `fields = [("DMSwarmPIC_coor", coordsdim, ScalarType)]`.
+        All fields must have the same number of points. For more
+        information see https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/DMSWARM/DMSWARM.html
+    :return: the immersed DMSwarm
+
+    .. note::
+
+        The created DMSwarm uses the communicator of the input DMPlex.
+    """
+
+    coords = np.asarray(coords, dtype=np.double)
+
+    # Create a DMSWARM
+    swarm = PETSc.DMSwarm().create(comm=plex.comm)
+
+    # Set swarm DM dimension to match DMPlex dimension
+    # NB: Unlike a DMPlex, this does not correspond to the topological
+    #     dimension of a mesh (which would be 0). In all PETSc examples
+    #     the dimension of the DMSwarm is set to match that of the
+    #     DMPlex used with swarm.setCellDM
+    swarm.setDimension(plex.getDimension())
+
+    # Set coordinates dimension
+    npoints, coordsdim = coords.shape
+    swarm.setCoordinateDim(coordsdim)
+
+    if coordsdim == 1:
+        raise NotImplementedError("1D DMSwarm not yet supported")
+
+    # Link to DMPlex cells information for when swarm.migrate() is used
+    swarm.setCellDM(plex)
+
+    # Set to Particle In Cell (PIC) type
+    swarm.setType(PETSc.DMSwarm.Type.PIC)
+
+    # Register any fields
+    fields = kwargs.get("fields", [])
+    for i in range(len(fields)):
+        swarm.registerField(fields[i][0], fields[i][1], dtype=fields[i][2])
+    swarm.finalizeFieldRegister()
+
+    # Note that no new fields can now be associated with the DMSWARM.
+
+    # Add point coordinates - note we set redundant mode to False since
+    # multiple MPI may be given the same list of coordinates. The cell
+    # DM (`plex`) will then attempt to locate the coordinates within its
+    # rank-local sub domain and disregard those which are outside it.
+    # See https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/DMSWARM/DMSwarmSetPointCoordinates.html
+    # for more information. The result is that all DMPlex cells,
+    # including ghost cells on distributed meshes, have the relevent PIC
+    # coordinates associated with them. The DMPlex cell id associated
+    # with each PIC in the DMSwarm is accessed with the `DMSwarm_cellid`
+    # field.
+    # NOTE This does not work correctly for periodic meshes for unknown
+    # reasons.
```

Comment:
This doesn't work for _known_ reasons, and you handle those cases higher up, so remove this comment.

Reply:
fixed